File: frontend/src/domain/usecases/ProductUseCase.ts
import { Product } from '../entities/Product';
import { ProductRepository, ProductFilter } from '../repositories/ProductRepository';

// Casos de uso para produtos
export interface ProductUseCase {
  getProducts(filters?: ProductFilter): Promise<Product[]>;
  getProduct(id: number): Promise<Product>;
  createProduct(product: Omit<Product, 'id'>): Promise<Product>;
  updateProduct(id: number, product: Partial<Product>): Promise<Product>;
  deleteProduct(id: number): Promise<void>;
  searchProducts(query: string): Promise<Product[]>;
  filterProductsByCategory(category: string): Promise<Product[]>;
}

// Implementação dos casos de uso
export class ProductUseCaseImpl implements ProductUseCase {
  constructor(private productRepository: ProductRepository) {}

  async getProducts(filters?: ProductFilter): Promise<Product[]> {
    try {
      return await this.productRepository.getProducts(filters);
    } catch (error) {
      console.error('Erro ao buscar produtos:', error);
      throw new Error('Falha ao carregar produtos');
    }
  }

  async getProduct(id: number): Promise<Product> {
    try {
      return await this.productRepository.getProduct(id);
    } catch (error) {
      console.error('Erro ao buscar produto:', error);
      throw new Error('Produto não encontrado');
    }
  }

  async createProduct(product: Omit<Product, 'id'>): Promise<Product> {
    try {
      // Validações de negócio
      if (!product.name || product.name.trim().length === 0) {
        throw new Error('Nome do produto é obrigatório');
      }
      if (product.price <= 0) {
        throw new Error('Preço deve ser maior que zero');
      }

      return await this.productRepository.createProduct(product);
    } catch (error) {
      console.error('Erro ao criar produto:', error);
      throw error;
    }
  }

  async updateProduct(id: number, product: Partial<Product>): Promise<Product> {
    try {
      // Validações de negócio
      if (product.name !== undefined && product.name.trim().length === 0) {
        throw new Error('Nome do produto é obrigatório');
      }
      if (product.price !== undefined && product.price <= 0) {
        throw new Error('Preço deve ser maior que zero');
      }

      return await this.productRepository.updateProduct(id, product);
    } catch (error) {
      console.error('Erro ao atualizar produto:', error);
      throw error;
    }
  }

  async deleteProduct(id: number): Promise<void> {
    try {
      await this.productRepository.deleteProduct(id);
    } catch (error) {
      console.error('Erro ao deletar produto:', error);
      throw new Error('Falha ao deletar produto');
    }
  }

  async searchProducts(query: string): Promise<Product[]> {
    try {
      const products = await this.productRepository.getProducts();
      return products.filter(product => 
        product.name.toLowerCase().includes(query.toLowerCase()) ||
        product.description.toLowerCase().includes(query.toLowerCase())
      );
    } catch (error) {
      console.error('Erro ao buscar produtos:', error);
      throw new Error('Falha na busca de produtos');
    }
  }

  async filterProductsByCategory(category: string): Promise<Product[]> {
    try {
      return await this.productRepository.getProducts({ category });
    } catch (error) {
      console.error('Erro ao filtrar produtos por categoria:', error);
      throw new Error('Falha ao filtrar produtos');
    }
  }
} 

File: backend/internal/domain/usecases/product_usecase.go
package usecases

import (
	"catalogo-produtos/backend/internal/domain/entities"
	"catalogo-produtos/backend/internal/domain/repositories"
	"errors"
)

// ProductUseCase define os casos de uso para produtos
type ProductUseCase interface {
	CreateProduct(name, image, description string, price float64, categoryID uint) (*entities.Product, error)
	GetProduct(id uint) (*entities.Product, error)
	GetProducts(filters *repositories.ProductFilter) ([]entities.Product, error)
	UpdateProduct(id uint, name, image, description string, price float64, categoryID uint) (*entities.Product, error)
	DeleteProduct(id uint) error
}

// productUseCase implementa ProductUseCase
type productUseCase struct {
	productRepo  repositories.ProductRepository
	categoryRepo repositories.CategoryRepository
}

// NewProductUseCase cria uma nova instância de ProductUseCase
func NewProductUseCase(productRepo repositories.ProductRepository, categoryRepo repositories.CategoryRepository) ProductUseCase {
	return &productUseCase{
		productRepo:  productRepo,
		categoryRepo: categoryRepo,
	}
}

// CreateProduct cria um novo produto
func (uc *productUseCase) CreateProduct(name, image, description string, price float64, categoryID uint) (*entities.Product, error) {
	// Validar se a categoria existe
	_, err := uc.categoryRepo.GetByID(categoryID)
	if err != nil {
		return nil, errors.New("categoria não encontrada")
	}

	// Validar preço
	if price <= 0 {
		return nil, errors.New("preço deve ser maior que zero")
	}

	// Validar nome
	if name == "" {
		return nil, errors.New("nome é obrigatório")
	}

	product := &entities.Product{
		Name:        name,
		Image:       image,
		Price:       price,
		CategoryID:  categoryID,
		Description: description,
	}

	err = uc.productRepo.Create(product)
	if err != nil {
		return nil, err
	}

	return product, nil
}

// GetProduct busca um produto por ID
func (uc *productUseCase) GetProduct(id uint) (*entities.Product, error) {
	product, err := uc.productRepo.GetByID(id)
	if err != nil {
		return nil, errors.New("produto não encontrado")
	}
	return product, nil
}

// GetProducts busca produtos com filtros
func (uc *productUseCase) GetProducts(filters *repositories.ProductFilter) ([]entities.Product, error) {
	products, err := uc.productRepo.GetAll(filters)
	if err != nil {
		return nil, err
	}
	return products, nil
}

// UpdateProduct atualiza um produto
func (uc *productUseCase) UpdateProduct(id uint, name, image, description string, price float64, categoryID uint) (*entities.Product, error) {
	// Buscar produto existente
	product, err := uc.productRepo.GetByID(id)
	if err != nil {
		return nil, errors.New("produto não encontrado")
	}

	// Validar se a categoria existe
	_, err = uc.categoryRepo.GetByID(categoryID)
	if err != nil {
		return nil, errors.New("categoria não encontrada")
	}

	// Validar preço
	if price <= 0 {
		return nil, errors.New("preço deve ser maior que zero")
	}

	// Validar nome
	if name == "" {
		return nil, errors.New("nome é obrigatório")
	}

	// Atualizar campos
	product.Name = name
	product.Image = image
	product.Price = price
	product.CategoryID = categoryID
	product.Description = description

	err = uc.productRepo.Update(product)
	if err != nil {
		return nil, err
	}

	return product, nil
}

// DeleteProduct remove um produto
func (uc *productUseCase) DeleteProduct(id uint) error {
	// Verificar se o produto existe
	_, err := uc.productRepo.GetByID(id)
	if err != nil {
		return errors.New("produto não encontrado")
	}

	return uc.productRepo.Delete(id)
}
